causallib/time_varying/GFormula.py

```diff
@@ -0,0 +1,218 @@
+                                    ) -> pd.DataFrame:
+        """
+              Returns individual estimated curves for each subject row in X/a/t
+
+              Args:
+                  X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                  a (pd.Series): Treatment assignment of size (num_subjects,).
+                  t (pd.Series): Followup durations, size (num_subjects,).
+                  y: NOT USED (for API compatibility only).
+                  timeline_start (int): Common start time-step. If provided, will generate survival curves starting
+                                        from 'timeline_start' for all patients.
+                                        If None, will predict from first observed event (t.min()).
+                  timeline_end (int): Common end time-step. If provided, will generate survival curves up to 'timeline_end'
+                                      for all patients.
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
+              """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def estimate_population_outcome(self,
+                                    X: pd.DataFrame,
+                                    a: pd.Series,
+                                    t: pd.Series,
+                                    y: Optional[Any] = None,
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None
+                                    ) -> pd.DataFrame:
+        """
+              Returns individual estimated curves for each subject row in X/a/t
+
+              Args:
+                  X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                  a (pd.Series): Treatment assignment of size (num_subjects,).
+                  t (pd.Series): Followup durations, size (num_subjects,).
+                  y: NOT USED (for API compatibility only).
+                  timeline_start (int): Common start time-step. If provided, will generate survival curves starting
+                                        from 'timeline_start' for all patients.
+                                        If None, will predict from first observed event (t.min()).
+                  timeline_end (int): Common end time-step. If provided, will generate survival curves up to 'timeline_end'
+                                      for all patients.
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
+              """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def apply_noise(self):
```

Comment:
1. probably should not be in `Base` since it's a specific implementation of the monte carlo version of G-formula.
2. should probably be private method: `_apply_noise`

Reply:
I've made this method abstract in base class, so that we could have specific implementation for each sub classes.
I'm okay if you want me to implement as a private method.

Reply:
yes lets move it to the specific implementation of GFormula